File: sdk/android/tracelet-sdk/src/main/kotlin/com/ikolvi/tracelet/sdk/location/DeadReckoningEngine.kt
package com.ikolvi.tracelet.sdk.location

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.os.Handler
import android.os.HandlerThread
import android.os.Looper
import android.util.Log
import com.ikolvi.tracelet.sdk.ConfigManager
import kotlin.math.*

/**
 * Inertial dead reckoning engine for GPS-denied environments.
 *
 * Uses accelerometer + magnetometer to estimate position when GPS is lost.
 * Implements Pedestrian Dead Reckoning (PDR) with step detection and heading
 * estimation. For vehicle/cycling modes, uses direct acceleration integration
 * with a high-pass filter.
 *
 * Accuracy degrades linearly with elapsed time:
 * - Pedestrian: base 5m + 1m/s
 * - Vehicle: base 10m + 3m/s
 *
 * Lifecycle:
 * 1. [activate] — start sensors, begin estimating from last known GPS position
 * 2. Engine emits estimated positions via [onEstimatedLocation] every ~1s
 * 3. [deactivate] — stop sensors, clean up
 * 4. Auto-stops after [ConfigManager.getDeadReckoningMaxDuration] seconds
 */
class DeadReckoningEngine(
    private val context: Context,
    private val config: ConfigManager,
) {
    companion object {
        private const val TAG = "DeadReckoningEngine"

        /** Meters per degree of latitude (approximate). */
        private const val METERS_PER_DEG_LAT = 111_139.0

        /** Sensor update interval: 50 Hz (20ms) for IMU fusion. */
        private const val SENSOR_DELAY_US = 20_000

        /** Minimum acceleration peak-to-trough for a valid step (m/s²). */
        private const val STEP_THRESHOLD = 2.0

        /** Minimum time between steps (ms) — prevents double-counting. */
        private const val MIN_STEP_INTERVAL_MS = 250L

        /** Maximum time between steps (ms) — if exceeded, reset peak detector. */
        private const val MAX_STEP_INTERVAL_MS = 2000L

        /** High-pass filter coefficient for vehicle acceleration integration. */
        private const val HIGH_PASS_ALPHA = 0.8

        /** Location emission interval (ms). */
        private const val EMIT_INTERVAL_MS = 1000L
    }

    // State
    @Volatile
    var isActive: Boolean = false
        private set
    private var activationTimeMs: Long = 0

    // Position
    private var currentLat: Double = 0.0
    private var currentLng: Double = 0.0
    private var currentAltitude: Double = 0.0
    private var currentHeading: Double = 0.0 // degrees from true north

    // Step detection (PDR)
    private var lastAccelMagnitude: Double = 0.0
    private var accelPeak: Double = 0.0
    private var accelTrough: Double = Double.MAX_VALUE
    private var lastStepTimeMs: Long = 0
    private var stepCount: Int = 0
    private var isAscending: Boolean = false

    // Vehicle mode integration
    private var velocityX: Double = 0.0
    private var velocityY: Double = 0.0
    private var lastSensorTimeNs: Long = 0
    private var filteredAccelX: Double = 0.0
    private var filteredAccelY: Double = 0.0

    // Heading from rotation matrix
    private val gravity = FloatArray(3)
    private val geomagnetic = FloatArray(3)
    private var hasGravity = false
    private var hasMagnetic = false

    // Sensors
    private var sensorManager: SensorManager? = null
    private var sensorThread: HandlerThread? = null
    private var sensorHandler: Handler? = null
    private val accelListener = AccelListener()
    private val magnetListener = MagnetListener()

    // Main thread for callbacks and timers
    private val mainHandler = Handler(Looper.getMainLooper())
    private var emitRunnable: Runnable? = null
    private var maxDurationRunnable: Runnable? = null

    // Activity type for algorithm selection
    private var activityType: String = "unknown"

    /** Callback: emits estimated location data. */
    var onEstimatedLocation: ((Map<String, Any?>) -> Unit)? = null

    /** Callback: notifies when dead reckoning auto-stops (max duration). */
    var onDeactivated: (() -> Unit)? = null

    // =========================================================================
    // Public API
    // =========================================================================

    /**
     * Activate dead reckoning from the last known GPS position.
     *
     * @param lat Last GPS latitude
     * @param lng Last GPS longitude
     * @param altitude Last GPS altitude
     * @param heading Last GPS heading (degrees)
     * @param activity Current detected activity type (e.g. "walking", "in_vehicle")
     */
    fun activate(
        lat: Double,
        lng: Double,
        altitude: Double,
        heading: Double,
        activity: String,
    ) {
        if (isActive) return

        currentLat = lat
        currentLng = lng
        currentAltitude = altitude
        currentHeading = if (heading >= 0) heading else 0.0
        activityType = activity
        activationTimeMs = System.currentTimeMillis()
        isActive = true

        // Reset state
        stepCount = 0
        lastStepTimeMs = 0
        accelPeak = 0.0
        accelTrough = Double.MAX_VALUE
        isAscending = false
        velocityX = 0.0
        velocityY = 0.0
        lastSensorTimeNs = 0
        filteredAccelX = 0.0
        filteredAccelY = 0.0
        hasGravity = false
        hasMagnetic = false

        Log.d(TAG, "Activated at ($lat, $lng), heading=$heading, activity=$activity")

        startSensors()
        startEmitTimer()
        startMaxDurationTimer()
    }

    /** Deactivate dead reckoning and release all sensor resources. */
    fun deactivate() {
        if (!isActive) return
        isActive = false

        stopSensors()
        stopEmitTimer()
        stopMaxDurationTimer()

        Log.d(TAG, "Deactivated after ${getElapsedSeconds()}s, $stepCount steps")
    }

    /** Returns the current DR state, or null if not active. */
    fun getState(): Map<String, Any?>? {
        if (!isActive) return null
        val elapsed = getElapsedSeconds()
        return mapOf(
            "active" to true,
            "elapsed" to elapsed,
            "estimatedAccuracy" to computeAccuracy(elapsed),
            "latitude" to currentLat,
            "longitude" to currentLng,
            "heading" to currentHeading,
            "stepCount" to stepCount,
            "activityType" to activityType,
        )
    }

    // =========================================================================
    // Sensor management
    // =========================================================================

    private fun startSensors() {
        sensorThread = HandlerThread("DeadReckoningSensors").apply { start() }
        sensorHandler = Handler(sensorThread!!.looper)

        val sm = context.getSystemService(Context.SENSOR_SERVICE) as? SensorManager ?: return
        sensorManager = sm

        // Accelerometer for step detection / acceleration integration
        sm.getDefaultSensor(Sensor.TYPE_LINEAR_ACCELERATION)?.let { sensor ->
            sm.registerListener(accelListener, sensor, SENSOR_DELAY_US, sensorHandler)
        } ?: run {
            // Fallback: use raw accelerometer (includes gravity)
            sm.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)?.let { sensor ->
                sm.registerListener(accelListener, sensor, SENSOR_DELAY_US, sensorHandler)
            }
        }

        // Magnetometer for heading
        sm.getDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD)?.let { sensor ->
            sm.registerListener(magnetListener, sensor, SENSOR_DELAY_US, sensorHandler)
        }

        // Gravity sensor for rotation matrix (separate from linear acceleration)
        sm.getDefaultSensor(Sensor.TYPE_GRAVITY)?.let { sensor ->
            sm.registerListener(gravityListener, sensor, SENSOR_DELAY_US, sensorHandler)
        }
    }

    private fun stopSensors() {
        sensorManager?.let { sm ->
            sm.unregisterListener(accelListener)
            sm.unregisterListener(magnetListener)
            sm.unregisterListener(gravityListener)
        }
        sensorManager = null

        sensorThread?.quitSafely()
        sensorThread = null
        sensorHandler = null
    }

    // =========================================================================
    // Sensor listeners
    // =========================================================================

    private inner class AccelListener : SensorEventListener {
        override fun onSensorChanged(event: SensorEvent) {
            if (!isActive) return

            val x = event.values[0].toDouble()
            val y = event.values[1].toDouble()
            val z = event.values[2].toDouble()
            val magnitude = sqrt(x * x + y * y + z * z)

            if (isVehicleMode()) {
                processVehicleAcceleration(x, y, event.timestamp)
            } else {
                processStepDetection(magnitude)
            }
        }

        override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {}
    }

    private inner class MagnetListener : SensorEventListener {
        override fun onSensorChanged(event: SensorEvent) {
            if (!isActive) return
            geomagnetic[0] = event.values[0]
            geomagnetic[1] = event.values[1]
            geomagnetic[2] = event.values[2]
            hasMagnetic = true
            updateHeading()
        }

        override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {}
    }

    private val gravityListener = object : SensorEventListener {
        override fun onSensorChanged(event: SensorEvent) {
            if (!isActive) return
            gravity[0] = event.values[0]
            gravity[1] = event.values[1]
            gravity[2] = event.values[2]
            hasGravity = true
            updateHeading()
        }

        override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {}
    }

    // =========================================================================
    // Heading computation
    // =========================================================================

    private fun updateHeading() {
        if (!hasGravity || !hasMagnetic) return

        val rotationMatrix = FloatArray(9)
        val inclinationMatrix = FloatArray(9)
        if (!SensorManager.getRotationMatrix(rotationMatrix, inclinationMatrix, gravity, geomagnetic)) {
            return
        }

        val orientation = FloatArray(3)
        SensorManager.getOrientation(rotationMatrix, orientation)

        // orientation[0] = azimuth in radians (-π to π)
        var azimuthDeg = Math.toDegrees(orientation[0].toDouble())
        if (azimuthDeg < 0) azimuthDeg += 360.0
        currentHeading = azimuthDeg
    }

    // =========================================================================
    // Pedestrian Dead Reckoning (PDR) — step detection
    // =========================================================================

    /**
     * Step detection via peak-trough analysis on acceleration magnitude.
     * Uses Weinberg step length estimation: stepLength = K * (peak - trough)^0.25
     */
    private fun processStepDetection(magnitude: Double) {
        val now = System.currentTimeMillis()

        // Track peaks and troughs
        if (magnitude > lastAccelMagnitude) {
            if (!isAscending && lastAccelMagnitude < accelTrough) {
                accelTrough = lastAccelMagnitude
            }
            isAscending = true
        } else {
            if (isAscending && lastAccelMagnitude > accelPeak) {
                accelPeak = lastAccelMagnitude
            }
            isAscending = false

            // Check for step: peak-to-trough exceeds threshold
            val diff = accelPeak - accelTrough
            if (diff > STEP_THRESHOLD) {
                val timeSinceLastStep = now - lastStepTimeMs
                if (timeSinceLastStep >= MIN_STEP_INTERVAL_MS) {
                    // Valid step detected
                    val stepLength = estimateStepLength(diff)
                    advancePosition(stepLength)
                    stepCount++
                    lastStepTimeMs = now

                    // Reset peak/trough for next step
                    accelPeak = 0.0
                    accelTrough = Double.MAX_VALUE
                }
            }
        }

        // Reset peak detector if no step for too long (stopped walking?)
        if (lastStepTimeMs > 0 && (now - lastStepTimeMs) > MAX_STEP_INTERVAL_MS) {
            accelPeak = 0.0
            accelTrough = Double.MAX_VALUE
        }

        lastAccelMagnitude = magnitude
    }

    /**
     * Estimate step length using the Weinberg formula.
     * stepLength = K * (accelPeak - accelTrough)^0.25
     * K ≈ 0.7 for average adult walking
     */
    private fun estimateStepLength(peakToDiff: Double): Double {
        return 0.7 * peakToDiff.pow(0.25)
    }

    // =========================================================================
    // Vehicle mode — acceleration integration
    // =========================================================================

    /**
     * For vehicle/cycling: integrate linear acceleration to estimate displacement.
     * Uses a high-pass filter to reduce drift from sensor bias.
     */
    private fun processVehicleAcceleration(ax: Double, ay: Double, timestampNs: Long) {
        if (lastSensorTimeNs == 0L) {
            lastSensorTimeNs = timestampNs
            return
        }

        val dt = (timestampNs - lastSensorTimeNs) / 1_000_000_000.0
        lastSensorTimeNs = timestampNs

        if (dt <= 0 || dt > 0.5) return // Skip invalid intervals

        // High-pass filter to remove low-frequency drift
        filteredAccelX = HIGH_PASS_ALPHA * (filteredAccelX + ax)
        filteredAccelY = HIGH_PASS_ALPHA * (filteredAccelY + ay)

        // Integrate acceleration → velocity
        velocityX += filteredAccelX * dt
        velocityY += filteredAccelY * dt

        // Apply velocity damping to prevent runaway drift
        velocityX *= 0.98
        velocityY *= 0.98

        // Integrate velocity → displacement (in device frame)
        val dx = velocityX * dt
        val dy = velocityY * dt

        // Transform from device frame to world frame using heading
        val headingRad = Math.toRadians(currentHeading)
        val worldDx = dx * cos(headingRad) - dy * sin(headingRad)
        val worldDy = dx * sin(headingRad) + dy * cos(headingRad)

        // Convert meters to degrees
        val metersPerDegLng = METERS_PER_DEG_LAT * cos(Math.toRadians(currentLat))
        currentLat += worldDy / METERS_PER_DEG_LAT
        currentLng += worldDx / metersPerDegLng
    }

    // =========================================================================
    // Position advancement (PDR)
    // =========================================================================

    /** Advance position by [stepLength] meters in the current heading direction. */
    private fun advancePosition(stepLength: Double) {
        val headingRad = Math.toRadians(currentHeading)

        // North displacement = stepLength * cos(heading)
        // East displacement = stepLength * sin(heading)
        val metersPerDegLng = METERS_PER_DEG_LAT * cos(Math.toRadians(currentLat))

        currentLat += (stepLength * cos(headingRad)) / METERS_PER_DEG_LAT
        currentLng += (stepLength * sin(headingRad)) / metersPerDegLng
    }

    // =========================================================================
    // Location emission
    // =========================================================================

    private fun startEmitTimer() {
        emitRunnable = object : Runnable {
            override fun run() {
                if (!isActive) return
                emitLocation()
                mainHandler.postDelayed(this, EMIT_INTERVAL_MS)
            }
        }
        mainHandler.postDelayed(emitRunnable!!, EMIT_INTERVAL_MS)
    }

    private fun stopEmitTimer() {
        emitRunnable?.let { mainHandler.removeCallbacks(it) }
        emitRunnable = null
    }

    private fun emitLocation() {
        val elapsed = getElapsedSeconds()
        val accuracy = computeAccuracy(elapsed)

        val location = mapOf<String, Any?>(
            "latitude" to currentLat,
            "longitude" to currentLng,
            "altitude" to currentAltitude,
            "heading" to currentHeading,
            "accuracy" to accuracy,
            "speed" to estimateSpeed(),
            "elapsed" to elapsed,
            "isDeadReckoned" to true,
        )

        onEstimatedLocation?.invoke(location)
    }

    // =========================================================================
    // Max duration auto-stop
    // =========================================================================

    private fun startMaxDurationTimer() {
        val maxDurationMs = config.getDeadReckoningMaxDuration() * 1000L
        maxDurationRunnable = Runnable {
            Log.d(TAG, "Max duration reached (${config.getDeadReckoningMaxDuration()}s)")
            deactivate()
            onDeactivated?.invoke()
        }
        mainHandler.postDelayed(maxDurationRunnable!!, maxDurationMs)
    }

    private fun stopMaxDurationTimer() {
        maxDurationRunnable?.let { mainHandler.removeCallbacks(it) }
        maxDurationRunnable = null
    }

    // =========================================================================
    // Helpers
    // =========================================================================

    private fun getElapsedSeconds(): Int {
        return ((System.currentTimeMillis() - activationTimeMs) / 1000).toInt()
    }

    /** Accuracy degrades linearly with time. */
    private fun computeAccuracy(elapsedSeconds: Int): Double {
        return if (isVehicleMode()) {
            10.0 + elapsedSeconds * 3.0
        } else {
            5.0 + elapsedSeconds * 1.0
        }
    }

    /** Estimate current speed from step frequency or velocity magnitude. */
    private fun estimateSpeed(): Double {
        if (isVehicleMode()) {
            return sqrt(velocityX * velocityX + velocityY * velocityY)
        }
        // Pedestrian: estimate from recent step timing
        val timeSinceLastStep = System.currentTimeMillis() - lastStepTimeMs
        return if (lastStepTimeMs > 0 && timeSinceLastStep < MAX_STEP_INTERVAL_MS) {
            // Average walking speed ~1.4 m/s
            0.7 * STEP_THRESHOLD.pow(0.25) * 1000.0 / timeSinceLastStep.coerceAtLeast(250)
        } else {
            0.0
        }
    }

    private fun isVehicleMode(): Boolean {
        return activityType in listOf("in_vehicle", "on_bicycle")
    }
}

File: sdk/ios/Sources/TraceletSDK/location/DeadReckoningEngine.swift
import CoreMotion
import Foundation

/// Inertial dead reckoning engine for GPS-denied environments.
///
/// Uses accelerometer + magnetometer (via `CMDeviceMotion`) to estimate
/// position when GPS is lost. Implements Pedestrian Dead Reckoning (PDR)
/// with step detection and heading estimation. For vehicle/cycling modes,
/// uses direct acceleration integration with a high-pass filter.
///
/// Accuracy degrades linearly with elapsed time:
/// - Pedestrian: base 5m + 1m/s
/// - Vehicle: base 10m + 3m/s
///
/// Lifecycle:
/// 1. ``activate(lat:lng:altitude:heading:activity:)`` — start sensors
/// 2. Engine emits estimated positions via ``onEstimatedLocation`` every ~1s
/// 3. ``deactivate()`` — stop sensors, clean up
/// 4. Auto-stops after configured max duration
public final class DeadReckoningEngine {
    /// Meters per degree of latitude (approximate).
    private static let metersPerDegLat: Double = 111_139.0

    /// Minimum acceleration peak-to-trough for a valid step (g-force).
    private static let stepThreshold: Double = 0.2

    /// Minimum time between steps — prevents double-counting.
    private static let minStepInterval: TimeInterval = 0.25

    /// Maximum time between steps — if exceeded, reset peak detector.
    private static let maxStepInterval: TimeInterval = 2.0

    /// High-pass filter coefficient for vehicle acceleration integration.
    private static let highPassAlpha: Double = 0.8

    /// Location emission interval.
    private static let emitInterval: TimeInterval = 1.0

    // State
    public private(set) var isActive = false
    private var activationDate: Date?

    // Position
    private var currentLat: Double = 0
    private var currentLng: Double = 0
    private var currentAltitude: Double = 0
    private var currentHeading: Double = 0 // degrees from true north

    // Step detection (PDR)
    private var lastAccelMagnitude: Double = 0
    private var accelPeak: Double = 0
    private var accelTrough: Double = .greatestFiniteMagnitude
    private var lastStepDate: Date?
    private var stepCount: Int = 0
    private var isAscending = false

    // Vehicle mode integration
    private var velocityX: Double = 0
    private var velocityY: Double = 0
    private var lastSensorTime: TimeInterval = 0
    private var filteredAccelX: Double = 0
    private var filteredAccelY: Double = 0

    // Sensors
    private let motionManager = CMMotionManager()
    private let sensorQueue: OperationQueue = {
        let q = OperationQueue()
        q.name = "com.tracelet.deadreckoning"
        q.qualityOfService = .userInitiated
        return q
    }()

    // Timers
    private var emitTimer: Timer?
    private var maxDurationTimer: Timer?

    // Config
    private let configManager: ConfigManager

    // Activity type for algorithm selection
    private var activityType: String = "unknown"

    /// Callback: emits estimated location data.
    public var onEstimatedLocation: (([String: Any]) -> Void)?

    /// Callback: notifies when dead reckoning auto-stops (max duration).
    public var onDeactivated: (() -> Void)?

    public init(configManager: ConfigManager) {
        self.configManager = configManager
        self.activationDate = nil
        self.lastStepDate = nil
    }

    // MARK: - Public API

    /// Activate dead reckoning from the last known GPS position.
    public func activate(
        lat: Double,
        lng: Double,
        altitude: Double,
        heading: Double,
        activity: String
    ) {
        guard !isActive else { return }

        currentLat = lat
        currentLng = lng
        currentAltitude = altitude
        currentHeading = heading >= 0 ? heading : 0
        activityType = activity
        activationDate = Date()
        isActive = true

        // Reset state
        stepCount = 0
        lastStepDate = nil
        accelPeak = 0
        accelTrough = .greatestFiniteMagnitude
        isAscending = false
        velocityX = 0
        velocityY = 0
        lastSensorTime = 0
        filteredAccelX = 0
        filteredAccelY = 0
        lastAccelMagnitude = 0

        TraceletLog.debug("[Tracelet] DeadReckoning activated at (\(lat), \(lng)), heading=\(heading), activity=\(activity)")

        startSensors()
        startEmitTimer()
        startMaxDurationTimer()
    }

    /// Deactivate dead reckoning and release all sensor resources.
    public func deactivate() {
        guard isActive else { return }
        isActive = false

        stopSensors()
        stopEmitTimer()
        stopMaxDurationTimer()

        TraceletLog.debug("[Tracelet] DeadReckoning deactivated after \(getElapsedSeconds())s, \(stepCount) steps")
    }

    /// Returns the current DR state, or nil if not active.
    public func getState() -> [String: Any]? {
        guard isActive else { return nil }
        let elapsed = getElapsedSeconds()
        return [
            "active": true,
            "elapsed": elapsed,
            "estimatedAccuracy": computeAccuracy(elapsed),
            "latitude": currentLat,
            "longitude": currentLng,
            "heading": currentHeading,
            "stepCount": stepCount,
            "activityType": activityType,
        ]
    }

    // MARK: - Sensor Management

    private func startSensors() {
        // Use CMDeviceMotion when available (fused accelerometer + gyroscope + magnetometer)
        if motionManager.isDeviceMotionAvailable {
            motionManager.deviceMotionUpdateInterval = 1.0 / 50.0 // 50 Hz
            motionManager.startDeviceMotionUpdates(
                using: .xMagneticNorthZVertical,
                to: sensorQueue
            ) { [weak self] motion, error in
                guard let self = self, let motion = motion, error == nil else { return }
                self.processDeviceMotion(motion)
            }
        } else if motionManager.isAccelerometerAvailable {
            // Fallback: raw accelerometer only (no heading)
            motionManager.accelerometerUpdateInterval = 1.0 / 50.0
            motionManager.startAccelerometerUpdates(to: sensorQueue) { [weak self] data, error in
                guard let self = self, let data = data, error == nil else { return }
                let acc = data.acceleration
                let magnitude = sqrt(acc.x * acc.x + acc.y * acc.y + acc.z * acc.z) - 1.0
                self.processStepDetection(abs(magnitude))
            }
        }
    }

    private func stopSensors() {
        motionManager.stopDeviceMotionUpdates()
        motionManager.stopAccelerometerUpdates()
    }

    // MARK: - Device Motion Processing

    private func processDeviceMotion(_ motion: CMDeviceMotion) {
        guard isActive else { return }

        // Update heading from device motion attitude
        // heading = yaw angle in degrees (0 = magnetic north)
        var heading = motion.attitude.yaw * 180.0 / .pi
        if heading < 0 { heading += 360.0 }
        currentHeading = heading

        // User acceleration (gravity-subtracted)
        let ax = motion.userAcceleration.x
        let ay = motion.userAcceleration.y
        let az = motion.userAcceleration.z
        let magnitude = sqrt(ax * ax + ay * ay + az * az)

        if isVehicleMode() {
            processVehicleAcceleration(ax: ax, ay: ay, timestamp: motion.timestamp)
        } else {
            processStepDetection(magnitude)
        }
    }

    // MARK: - Pedestrian Dead Reckoning (PDR)

    /// Step detection via peak-trough analysis on acceleration magnitude.
    /// Uses Weinberg step length estimation.
    private func processStepDetection(_ magnitude: Double) {
        let now = Date()

        // Track peaks and troughs
        if magnitude > lastAccelMagnitude {
            if !isAscending && lastAccelMagnitude < accelTrough {
                accelTrough = lastAccelMagnitude
            }
            isAscending = true
        } else {
            if isAscending && lastAccelMagnitude > accelPeak {
                accelPeak = lastAccelMagnitude
            }
            isAscending = false

            // Check for step: peak-to-trough exceeds threshold
            let diff = accelPeak - accelTrough
            if diff > Self.stepThreshold {
                let timeSinceLastStep = lastStepDate.map { now.timeIntervalSince($0) } ?? Self.minStepInterval
                if timeSinceLastStep >= Self.minStepInterval {
                    let stepLength = estimateStepLength(diff)
                    advancePosition(stepLength)
                    stepCount += 1
                    lastStepDate = now

                    accelPeak = 0
                    accelTrough = .greatestFiniteMagnitude
                }
            }
        }

        // Reset peak detector if no step for too long
        if let lastStep = lastStepDate, now.timeIntervalSince(lastStep) > Self.maxStepInterval {
            accelPeak = 0
            accelTrough = .greatestFiniteMagnitude
        }

        lastAccelMagnitude = magnitude
    }

    /// Estimate step length using the Weinberg formula.
    private func estimateStepLength(_ peakToDiff: Double) -> Double {
        return 0.7 * pow(peakToDiff, 0.25)
    }

    // MARK: - Vehicle Mode

    private func processVehicleAcceleration(ax: Double, ay: Double, timestamp: TimeInterval) {
        if lastSensorTime == 0 {
            lastSensorTime = timestamp
            return
        }

        let dt = timestamp - lastSensorTime
        lastSensorTime = timestamp

        guard dt > 0, dt < 0.5 else { return }

        // High-pass filter
        filteredAccelX = Self.highPassAlpha * (filteredAccelX + ax)
        filteredAccelY = Self.highPassAlpha * (filteredAccelY + ay)

        // Integrate acceleration → velocity
        velocityX += filteredAccelX * dt
        velocityY += filteredAccelY * dt

        // Velocity damping
        velocityX *= 0.98
        velocityY *= 0.98

        // Integrate velocity → displacement
        let dx = velocityX * dt
        let dy = velocityY * dt

        // Transform to world frame using heading
        let headingRad = currentHeading * .pi / 180.0
        let worldDx = dx * cos(headingRad) - dy * sin(headingRad)
        let worldDy = dx * sin(headingRad) + dy * cos(headingRad)

        let metersPerDegLng = Self.metersPerDegLat * cos(currentLat * .pi / 180.0)
        currentLat += worldDy / Self.metersPerDegLat
        currentLng += worldDx / metersPerDegLng
    }

    // MARK: - Position Advancement

    private func advancePosition(_ stepLength: Double) {
        let headingRad = currentHeading * .pi / 180.0
        let metersPerDegLng = Self.metersPerDegLat * cos(currentLat * .pi / 180.0)

        currentLat += (stepLength * cos(headingRad)) / Self.metersPerDegLat
        currentLng += (stepLength * sin(headingRad)) / metersPerDegLng
    }

    // MARK: - Location Emission

    private func startEmitTimer() {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            self.emitTimer = Timer.scheduledTimer(
                withTimeInterval: Self.emitInterval,
                repeats: true
            ) { [weak self] _ in
                self?.emitLocation()
            }
        }
    }

    private func stopEmitTimer() {
        emitTimer?.invalidate()
        emitTimer = nil
    }

    private func emitLocation() {
        guard isActive else { return }
        let elapsed = getElapsedSeconds()
        let accuracy = computeAccuracy(elapsed)

        let location: [String: Any] = [
            "latitude": currentLat,
            "longitude": currentLng,
            "altitude": currentAltitude,
            "heading": currentHeading,
            "accuracy": accuracy,
            "speed": estimateSpeed(),
            "elapsed": elapsed,
            "isDeadReckoned": true,
        ]

        onEstimatedLocation?(location)
    }

    // MARK: - Max Duration

    private func startMaxDurationTimer() {
        let maxDuration = TimeInterval(configManager.getDeadReckoningMaxDuration())
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            self.maxDurationTimer = Timer.scheduledTimer(
                withTimeInterval: maxDuration,
                repeats: false
            ) { [weak self] _ in
                guard let self = self else { return }
                TraceletLog.debug("[Tracelet] DeadReckoning max duration reached (\(Int(maxDuration))s)")
                self.deactivate()
                self.onDeactivated?()
            }
        }
    }

    private func stopMaxDurationTimer() {
        maxDurationTimer?.invalidate()
        maxDurationTimer = nil
    }

    // MARK: - Helpers

    private func getElapsedSeconds() -> Int {
        guard let activation = activationDate else { return 0 }
        return Int(Date().timeIntervalSince(activation))
    }

    private func computeAccuracy(_ elapsedSeconds: Int) -> Double {
        if isVehicleMode() {
            return 10.0 + Double(elapsedSeconds) * 3.0
        }
        return 5.0 + Double(elapsedSeconds) * 1.0
    }

    private func estimateSpeed() -> Double {
        if isVehicleMode() {
            return sqrt(velocityX * velocityX + velocityY * velocityY)
        }
        guard let lastStep = lastStepDate else { return 0 }
        let timeSinceLastStep = Date().timeIntervalSince(lastStep)
        if timeSinceLastStep < Self.maxStepInterval {
            return 0.7 * pow(Self.stepThreshold, 0.25) * 1.0 / max(timeSinceLastStep, Self.minStepInterval)
        }
        return 0
    }

    private func isVehicleMode() -> Bool {
        return activityType == "in_vehicle" || activityType == "on_bicycle"
    }
}
